pkg/usagetracker/partition_handler.go

```diff
@@ -0,0 +1,972 @@
+
+	if p.cfg.SkipSnapshotLoadingAtStartup {
+		level.Warn(p.logger).Log("msg", "configured to skip snapshot loading at startup")
+		return 0, nil
+	}
+
+	offset, err := findEndOffset(ctx, p.snapshotsKafkaReader, snapshotsTopic, p.partitionID)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to find end offset for topic %s", snapshotsTopic)
+	}
+	if offset == 0 {
+		level.Warn(p.logger).Log("msg", "end offset is 0, no snapshots found", "topic", snapshotsTopic)
+		return 0, nil
+	}
+
+	// Start reading the last snapshot.
+	fetches := p.snapshotsKafkaReader.PollFetches(ctx)
+	if err := fetches.Err(); err != nil {
+		return 0, fmt.Errorf("failed to retrieve last snapshot: %w", err)
+	}
+	if fetches.Empty() {
+		return 0, fmt.Errorf("didn't get any snapshot records from topic %s", snapshotsTopic)
+	}
+	// Get the last record from the fetches.
+	record := fetches.Records()[len(fetches.Records())-1]
+	// Log it and continue
+	level.Info(p.logger).Log("msg", "got last snapshot record", "offset", record.Offset, "bytes", len(record.Value))
+
+	if time.Since(record.Timestamp) > p.cfg.IdleTimeout {
+		level.Warn(p.logger).Log("msg", "last snapshot record is too old, ignoring it", "offset", record.Offset, "timestamp", record.Timestamp, "idle_timeout", p.cfg.IdleTimeout)
+		return 0, nil
+	}
+
+	var snapshot usagetrackerpb.SnapshotRecord
+	if err := snapshot.Unmarshal(record.Value); err != nil {
+		return 0, errors.Wrapf(err, "failed to unmarshal snapshot record from topic %s", snapshotsTopic)
+	}
+
+	level.Info(p.logger).Log("msg", "loading snapshot record", "offset", record.Offset, "record_timestamp", record.Timestamp, "snapshot_timestamp", snapshot.Timestamp)
+	if err := p.loadAllSnapshotShards(ctx, snapshot.Filenames); err != nil {
+		return 0, errors.Wrapf(err, "failed to load snapshot")
+	}
+
+	p.loadedLastSnapshotTimestamp = time.UnixMilli(snapshot.Timestamp)
+	p.ignoreSnapshotEventsUntil = snapshot.LastSnapshotEventOffset
+
+	return snapshot.LastEventOffsetPublishedBeforeSnapshot, nil
+}
+
+func (p *partitionHandler) loadAllSnapshotShards(ctx context.Context, files []string) error {
+	t0 := time.Now()
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(nil)
+
+	type downloadedSnapshot struct {
+		filename string
+		data     [][]byte
+	}
+
+	// One is being downloaded, one is buffered, one is being processed.
+	downloaded := make(chan downloadedSnapshot, 1)
+	errs := make(chan error, 1)
+	go func() {
+		t0 := time.Now()
+		i := 0
+		for snapshot := range downloaded {
+			for j, data := range snapshot.data {
+				if err := p.store.loadSnapshot(data, time.Now()); err != nil {
+					errs <- errors.Wrapf(err, "failed to load snapshot data %d from file %q", j, snapshot.filename)
+					return
+				}
+			}
+			if i == 0 {
+				level.Info(p.logger).Log("msg", "loaded the first snapshot file", "filename", snapshot.filename, "shards", len(snapshot.data), "load_time", time.Since(t0))
+			}
+			i++
+		}
+		errs <- nil
+	}()
+
+	go func() {
```

Comment:
Mostly for code simplicity, instead of launching two goroutines (consumer and producer) and have the parent goroutine waiting, we can launch a single producer goroutine and consume in the parent goroutine.